src/RamTransforms.cpp

```diff
@@ -554,6 +554,190 @@ bool TupleIdTransformer::reorderOperations(RamProgram& program) {
     return changed;
 }
 
+bool HoistAggregateTransformer::hoistAggregate(RamProgram& program) {
+    // flag to determine whether the RAM program has changed
+    bool changed = false;
+
+    // hoist aggregate if doesn't depend on anything
+    visitDepthFirst(program, [&](const RamQuery& query) {
+        int currLevel = -1;
+        int oldLevel = -1;
+        bool hoist = false;
+        bool newIndex = false;
+
+        // Fields for new aggregate
+        AggregateFunction newFun;
+        const RamRelation* newRef;
+        RamExpression* newExp;
+        RamCondition* newCond;
+        std::vector<RamExpression*> newPattern;
+
+        // Tracking aggregates seen to determine whether to
+        // hoist or not
+        std::vector<int> aggIds;
+
+        // Removing an aggregate that can be hoisted if it exists
+        std::function<std::unique_ptr<RamNode>(std::unique_ptr<RamNode>)> aggRemover =
+                [&](std::unique_ptr<RamNode> node) -> std::unique_ptr<RamNode> {
+            // Checking that there exists an aggregate to hoist
+            // If so, remove the aggregate from the loop nest
+            if (const RamAggregate* agg = dynamic_cast<RamAggregate*>(node.get())) {
+                aggIds.push_back(agg->getTupleId());
+                // assuming that RamSearches have tupleIds 0, 1, 2, ...
+                if (!hoist && rla->getLevel(agg) < agg->getTupleId() - 1) {
+                    // If all searches between the rla->getLevel(agg) and agg
+                    // are aggregates, then we do not transform
+                    bool allAggragates = true;
+                    for (int i = rla->getLevel(agg) + 1; i <= agg->getTupleId(); i++) {
+                        if (!(std::find(aggIds.begin(), aggIds.end(), i) != aggIds.end())) {
+                            currLevel = i - 1;
+                            allAggragates = false;
+                            break;
+                        }
+                    }
+                    if (!allAggragates) {
```

Comment:
Aggragates -> Aggregates (and elsewhere)

Reply:
Oh, I didn't see this before I merged. You should have made it a review and requested changes  Spelling PR!

Reply:
This is also my bad - Bernhard wanted the logic of the transformer to be simplified a bit so I should have prevented merging. Will cover these two things in another PR.